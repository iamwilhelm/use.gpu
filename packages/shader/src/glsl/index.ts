import {
  loadModule,
  loadModuleWithCache,
  loadStaticModule,
  bindEntryPoint,
  defineConstants,
  makeModuleCache,

  bundleToAttribute,
  bundleToAttributes,

  glsl, float, int, uint,
} from './shader';

import {
  bindBundle,
  bindModule,
  bindingsToLinks,
  bindingToModule,
  resolveBindings,
} from './bind';

import {
  castTo,
  swizzleTo,
} from './cast';

import {
  chainTo,
} from './chain';

import {
  diffBy,
} from './diff';

import {
  linkBundle,
  linkModule,
  linkCode,
  getPreamble,
  setPreamble,
} from './link';

import {
  makeASTParser,
  compressAST,
  decompressAST,
  rewriteUsingAST,
} from './ast';

import {
  getHash
} from '../util/hash';

import {
  getBundleHash,
  getBundleKey,
} from '../util/bundle';

export {
  loadModule,
  loadModuleWithCache,
  loadStaticModule,
  bindEntryPoint,
  defineConstants,
  makeModuleCache,

  bundleToAttribute,
  bundleToAttributes,

  glsl, float, int, uint,
} from './shader';

export {
  bindBundle,
  bindModule,
  bindingsToLinks,
  bindingToModule,
  resolveBindings,
} from './bind';

export {
  castTo,
  swizzleTo,
} from './cast';

export {
  chainTo,
} from './chain';

export {
  diffBy,
} from './diff';

export {
  linkBundle,
  linkModule,
  linkCode,
  getPreamble,
  setPreamble,
} from './link';

export {
  makeASTParser,
  compressAST,
  decompressAST,
  rewriteUsingAST,
} from './ast';

export {
  getHash
} from '../util/hash';

export {
  getBundleHash,
  getBundleKey,
} from '../util/bundle';

export const GLSLLinker = {
  loadModule,
  loadModuleWithCache,
  loadStaticModule,
  bindEntryPoint,
  bundleToAttribute,
  bundleToAttributes,
  glsl, float, int, uint,

  defineConstants,

  linkBundle,
  linkModule,
  linkCode,
  getPreamble,
  setPreamble,

  bindBundle,
  bindModule,
  bindingsToLinks,
  bindingToModule,
  resolveBindings,

  castTo,
  chainTo,
  diffBy,
  swizzleTo,

  makeASTParser,
  compressAST,
  decompressAST,
  rewriteUsingAST,

  makeModuleCache,

  getHash,
  getBundleHash,
  getBundleKey,
};

export default GLSLLinker;