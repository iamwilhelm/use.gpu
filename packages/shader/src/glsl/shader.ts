import { Tree } from '@lezer/common';
import { ParsedModule, ParsedModuleCache, ShaderDefine } from './types';

import { makeLoadModule, makeLoadModuleWithCache } from '../util/shader';
import { makeBundleToAttribute, makeBundleToAttributes } from '../util/bundle';
import { decompressAST } from '../util/tree';

import { makeASTParser, compressAST } from './ast';
import { toTypeString, toTypeArgs } from './type';
import { parser } from './grammar/glsl';
import LRU from 'lru-cache';

export { loadStaticModule, loadVirtualModule } from '../util/shader';

// LRU cache for parsed shader code
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU<string, ParsedModule>({
  max: 100,
  ...options,
});

export const DEFAULT_CACHE = makeModuleCache();

// Parse GLSL code into lezer tree
export const parseShader = (code: string): Tree => parser.parse(code);

// Parse a code module into its in-memory representation
// (AST + symbol table)
export const loadModule = makeLoadModule(parseShader, makeASTParser, compressAST);

// Use cache to load modules
export const loadModuleWithCache = makeLoadModuleWithCache(loadModule, DEFAULT_CACHE);

// Make GLSL definitions
export const defineConstants = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (defs[k] !== false && defs[k] !== null) out.push(`#define ${k} ${defs[k]}`);
  return out.join("\n");
}

export const bundleToAttribute = makeBundleToAttribute(toTypeString, toTypeArgs);
export const bundleToAttributes = makeBundleToAttributes(toTypeString, toTypeArgs);
