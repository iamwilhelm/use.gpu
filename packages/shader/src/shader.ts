import { Tree, SyntaxNode } from '@lezer/common';
import { parser } from './grammar/glsl';

export const parseGLSL = (code: string): Tree => {
  const parsed = parser.parse(code);
  return parsed;
}
