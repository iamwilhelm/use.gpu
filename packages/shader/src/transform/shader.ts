import { ShaderLanguages, ShaderCompiler } from '@use-gpu/core/types';
import { ParsedModule, ParsedModuleCache, ShaderDefine } from '../types';
import { Tree, SyntaxNode } from '@lezer/common';

import { makeASTParser, compressAST, decompressAST, getProgramHash } from './ast';
import { parser } from '../grammar/glsl';
import LRU from 'lru-cache';

type LangDef = {
  glsl?: ShaderCompiler,
  cache?: any,
};

// Parse a code module into its in-memory representation
// (AST + symbol table)
export const loadModule = (
  code: string,
  name: string,
  compressed: boolean = false,
): ParsedModule => {
  if (code == null) throw new Error(`Shader code ${name} undefined`);
  let tree = parseGLSL(code);

  const parser = makeASTParser(code, tree);
  const table = parser.getSymbolTable();
  const shake = parser.getShakeTable(table);

  if (compressed) tree = decompressAST(compressAST(tree));

  return {name, code, tree, table, shake};
}

// Use cache to load modules
export const loadModuleWithCache = (
  code: string,
  name: string,
  cache: ParsedModuleCache | null = null,
): ParsedModule => {
  if (!cache) return loadModule(code, name, true);

  const hash = getProgramHash(code);
  const entry = cache.get(hash);
  if (entry) return entry;
  
  const module = loadModule(code, name, true);
  cache.set(hash, module);
  return module;
}

// Parse GLSL using lezer grammar
export const parseGLSL = (code: string): Tree => parser.parse(code);

// Make GLSL definitions
export const defineGLSL = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (defs[k] !== false && defs[k] !== null) out.push(`#define ${k} ${defs[k]}`);
  return out.join("\n");
}

// Make shader languages interface
export const makeShaderLanguages = (langs: LangDef[]): ShaderLanguages => {
  const out = {} as any;
  for (const {glsl} of langs) {
    if (glsl) {
      out.glsl = {
        compile: (code: string, stage: any) => (glsl as any)(code, stage, false),
        cache: makeModuleCache(),
      };
    }
  }
  return out;
};

// LRU cache for parsed shader code
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU({
  max: 100,
  ...options,
});
