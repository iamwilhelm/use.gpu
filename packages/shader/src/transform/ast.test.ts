import { GLSLModules } from '@use-gpu/glsl';
import { parseGLSL } from './shader';
import { makeASTParser, formatAST, rewriteUsingAST, compressAST, decompressAST, hasErrorNode } from './ast';
import { addASTSerializer } from '../test/snapshot';

addASTSerializer(expect);

describe('ast', () => {
  
  const makeGuardedParser = (code: any, tree: any): ReturnType<typeof makeASTParser> => {
    let errorNode = hasErrorNode(tree);
    if (errorNode) {
      for (let i = 0; i < 2; ++i) if (errorNode.parent) errorNode = errorNode.parent;
      console.log(formatAST(errorNode, code));
      throw new Error("Error in AST");
    }
    
    return makeASTParser(code, tree);
  }

  it('extracts test declarations', () => {
    const code = `
      float x, y;
      int a = 3, b = 1, c, d;
    `;

    const tree = parseGLSL(code);
    const {extractDeclarations} = makeGuardedParser(code, tree);

    const declarations = extractDeclarations();
    expect(declarations).toMatchSnapshot();
  });
  
  it('extracts test declarations with array', () => {
    const code = `
      const ivec2 QUAD[] = {
        ivec2(0, 0),
        ivec2(1, 0),
        ivec2(0, 1),
        ivec2(1, 1),
      };
    `;

    const tree = parseGLSL(code);
    const {extractDeclarations} = makeGuardedParser(code, tree);

    const declarations = extractDeclarations();
    expect(declarations).toMatchSnapshot();
  });
  
  it('extracts quad vertex imports', () => {
    const code = GLSLModules['instance/quad/vertex'];

    const tree = parseGLSL(code);
    const {extractImports} = makeGuardedParser(code, tree);

    const imports = extractImports();
    expect(imports).toMatchSnapshot();
  });
  
  it('extracts quad vertex functions', () => {
    const code = GLSLModules['instance/quad/vertex'];

    const tree = parseGLSL(code);
    const {extractFunctions} = makeGuardedParser(code, tree);

    const functions = extractFunctions();
    expect(functions).toMatchSnapshot();
  });

  it('extracts quad vertex declarations', () => {
    const code = GLSLModules['instance/quad/vertex'];

    const tree = parseGLSL(code);
    const {extractDeclarations} = makeGuardedParser(code, tree);

    const declarations = extractDeclarations();
    expect(declarations).toMatchSnapshot();
  });

  it('extracts quad fragment declarations', () => {
    const code = GLSLModules['instance/quad/fragment'];

    const tree = parseGLSL(code);
    const {extractDeclarations} = makeGuardedParser(code, tree);

    const declarations = extractDeclarations();
    expect(declarations).toMatchSnapshot();
  });

  it('extracts use view declarations', () => {
    const code = GLSLModules['use/view'];

    const tree = parseGLSL(code);
    const {extractDeclarations} = makeGuardedParser(code, tree);

    const declarations = extractDeclarations();
    expect(declarations).toMatchSnapshot();
  });

  it('extracts quad vertex symbol table', () => {
    const code = GLSLModules['instance/quad/vertex'];

    const tree = parseGLSL(code);
    const {extractSymbolTable} = makeGuardedParser(code, tree);

    const symbolTable = extractSymbolTable();
    expect(symbolTable).toMatchSnapshot();
  });

  it('extracts geometry quad symbol table', () => {
    const code = GLSLModules['geometry/quad'];

    const tree = parseGLSL(code);
    const {extractSymbolTable} = makeGuardedParser(code, tree);

    const symbolTable = extractSymbolTable();
    expect(symbolTable).toMatchSnapshot();
  });

  it('extracts use view symbol table', () => {
    const code = GLSLModules['use/view'];

    const tree = parseGLSL(code);
    const {extractSymbolTable} = makeGuardedParser(code, tree);

    const symbolTable = extractSymbolTable();
    expect(symbolTable).toMatchSnapshot();
  });

  it('extracts use types symbol table', () => {
    const code = GLSLModules['use/types'];

    const tree = parseGLSL(code);
    const {extractSymbolTable} = makeGuardedParser(code, tree);

    const symbolTable = extractSymbolTable();
    expect(symbolTable).toMatchSnapshot();
  });

  it('rewrites code using the AST', () => {
    const code = `
    float getValue(int index);
    void main() {
      float x = 3.0;
      float y = getValue(2);
    }
    `;

    const tree = parseGLSL(code);
    const rename = new Map<string, string>();
    rename.set('main', 'entryPoint');
    rename.set('getValue', '_zz_getValue');
    
    const output = rewriteUsingAST(code, tree, rename);
    expect(output).toMatchSnapshot();
  });

  it('rewrites code using the compressed AST', () => {
    const code = `
    float getValue(int index);
    void main() {
      float x = 3.0;
      float y = getValue(2);
    }
    `;

    const tree = parseGLSL(code);
    const rename = new Map<string, string>();
    rename.set('main', 'entryPoint');
    rename.set('getValue', '_zz_getValue');
    
    const compressed = compressAST(tree);
    const decompressed = decompressAST(compressed);
    expect(compressed).toMatchSnapshot();
    expect(decompressed).toMatchSnapshot();

    const output1 = rewriteUsingAST(code, tree, rename);
    const output2 = rewriteUsingAST(code, decompressed, rename);
    expect(output1).toEqual(output2);
  });
});
