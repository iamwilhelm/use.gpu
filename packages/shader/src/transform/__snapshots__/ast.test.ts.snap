// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ast extracts geometry quad symbol table 1`] = `
Object {
  "declarations": Array [
    Object {
      "exported": false,
      "symbols": Array [
        "QUAD",
      ],
      "variable": Object {
        "locals": Array [
          Object {
            "expr": "ivec2(0, 0)",
            "name": "QUAD",
          },
        ],
        "type": Object {
          "name": "ivec2",
          "qualifiers": Array [
            "const",
          ],
        },
      },
    },
  ],
  "externals": Array [],
  "functions": Array [
    Object {
      "exported": true,
      "prototype": Object {
        "name": "getQuadIndex",
        "parameters": Array [
          "int",
          "vertex",
        ],
        "type": Object {
          "name": "ivec2",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "getQuadIndex",
      ],
    },
    Object {
      "exported": true,
      "prototype": Object {
        "name": "getQuadUV",
        "parameters": Array [
          "int",
          "vertex",
        ],
        "type": Object {
          "name": "vec2",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "getQuadUV",
      ],
    },
  ],
  "hash": "wbun5hxc7b",
  "modules": Array [],
  "refs": Array [
    Object {
      "exported": true,
      "prototype": Object {
        "name": "getQuadIndex",
        "parameters": Array [
          "int",
          "vertex",
        ],
        "type": Object {
          "name": "ivec2",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "getQuadIndex",
      ],
    },
    Object {
      "exported": true,
      "prototype": Object {
        "name": "getQuadUV",
        "parameters": Array [
          "int",
          "vertex",
        ],
        "type": Object {
          "name": "vec2",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "getQuadUV",
      ],
    },
    Object {
      "exported": false,
      "symbols": Array [
        "QUAD",
      ],
      "variable": Object {
        "locals": Array [
          Object {
            "expr": "ivec2(0, 0)",
            "name": "QUAD",
          },
        ],
        "type": Object {
          "name": "ivec2",
          "qualifiers": Array [
            "const",
          ],
        },
      },
    },
  ],
  "symbols": Array [
    "getQuadIndex",
    "getQuadUV",
    "QUAD",
  ],
  "visibles": Array [
    "getQuadIndex",
    "getQuadUV",
  ],
}
`;

exports[`ast extracts quad fragment declarations 1`] = `
Array [
  Object {
    "exported": false,
    "symbols": Array [
      "fragColor",
    ],
    "variable": Object {
      "locals": Array [
        Object {
          "expr": null,
          "name": "fragColor",
        },
      ],
      "type": Object {
        "name": "vec4",
        "qualifiers": Array [
          "layout(location = 0)",
          "in",
        ],
      },
    },
  },
  Object {
    "exported": false,
    "symbols": Array [
      "fragUV",
    ],
    "variable": Object {
      "locals": Array [
        Object {
          "expr": null,
          "name": "fragUV",
        },
      ],
      "type": Object {
        "name": "vec2",
        "qualifiers": Array [
          "layout(location = 1)",
          "in",
        ],
      },
    },
  },
  Object {
    "exported": false,
    "symbols": Array [
      "outColor",
    ],
    "variable": Object {
      "locals": Array [
        Object {
          "expr": null,
          "name": "outColor",
        },
      ],
      "type": Object {
        "name": "vec4",
        "qualifiers": Array [
          "layout(location = 0)",
          "out",
        ],
      },
    },
  },
]
`;

exports[`ast extracts quad vertex declarations 1`] = `
Array [
  Object {
    "exported": false,
    "prototype": Object {
      "name": "getPosition",
      "parameters": Array [
        "int",
      ],
      "type": Object {
        "name": "vec4",
        "qualifiers": Array [],
      },
    },
    "symbols": Array [
      "getPosition",
    ],
  },
  Object {
    "exported": false,
    "prototype": Object {
      "name": "getSize",
      "parameters": Array [
        "int",
      ],
      "type": Object {
        "name": "float",
        "qualifiers": Array [],
      },
    },
    "symbols": Array [
      "getSize",
    ],
  },
]
`;

exports[`ast extracts quad vertex functions 1`] = `
Array [
  Object {
    "exported": true,
    "prototype": Object {
      "name": "getQuadVertex",
      "parameters": Array [
        "int",
        "vertexIndex",
      ],
      "type": Object {
        "name": "MeshVertex",
        "qualifiers": Array [],
      },
    },
    "symbols": Array [
      "getQuadVertex",
    ],
  },
]
`;

exports[`ast extracts quad vertex imports 1`] = `
Array [
  Object {
    "imports": Array [
      Object {
        "imported": "MeshVertex",
        "name": "MeshVertex",
      },
    ],
    "name": "use/types",
    "symbols": Array [
      "MeshVertex",
    ],
  },
  Object {
    "imports": Array [
      Object {
        "imported": "viewUniforms",
        "name": "viewUniforms",
      },
      Object {
        "imported": "worldToClip",
        "name": "worldToClip",
      },
    ],
    "name": "use/view",
    "symbols": Array [
      "viewUniforms",
      "worldToClip",
    ],
  },
  Object {
    "imports": Array [
      Object {
        "imported": "getQuadUV",
        "name": "getQuadUV",
      },
    ],
    "name": "geometry/quad",
    "symbols": Array [
      "getQuadUV",
    ],
  },
]
`;

exports[`ast extracts quad vertex symbol table 1`] = `
Object {
  "declarations": Array [
    Object {
      "exported": false,
      "prototype": Object {
        "name": "getPosition",
        "parameters": Array [
          "int",
        ],
        "type": Object {
          "name": "vec4",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "getPosition",
      ],
    },
    Object {
      "exported": false,
      "prototype": Object {
        "name": "getSize",
        "parameters": Array [
          "int",
        ],
        "type": Object {
          "name": "float",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "getSize",
      ],
    },
  ],
  "externals": Array [
    Object {
      "exported": false,
      "prototype": Object {
        "name": "getPosition",
        "parameters": Array [
          "int",
        ],
        "type": Object {
          "name": "vec4",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "getPosition",
      ],
    },
    Object {
      "exported": false,
      "prototype": Object {
        "name": "getSize",
        "parameters": Array [
          "int",
        ],
        "type": Object {
          "name": "float",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "getSize",
      ],
    },
  ],
  "functions": Array [
    Object {
      "exported": true,
      "prototype": Object {
        "name": "getQuadVertex",
        "parameters": Array [
          "int",
          "vertexIndex",
        ],
        "type": Object {
          "name": "MeshVertex",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "getQuadVertex",
      ],
    },
  ],
  "hash": "aassgu6e5l",
  "modules": Array [
    Object {
      "imports": Array [
        Object {
          "imported": "MeshVertex",
          "name": "MeshVertex",
        },
      ],
      "name": "use/types",
      "symbols": Array [
        "MeshVertex",
      ],
    },
    Object {
      "imports": Array [
        Object {
          "imported": "viewUniforms",
          "name": "viewUniforms",
        },
        Object {
          "imported": "worldToClip",
          "name": "worldToClip",
        },
      ],
      "name": "use/view",
      "symbols": Array [
        "viewUniforms",
        "worldToClip",
      ],
    },
    Object {
      "imports": Array [
        Object {
          "imported": "getQuadUV",
          "name": "getQuadUV",
        },
      ],
      "name": "geometry/quad",
      "symbols": Array [
        "getQuadUV",
      ],
    },
  ],
  "refs": Array [
    Object {
      "imports": Array [
        Object {
          "imported": "MeshVertex",
          "name": "MeshVertex",
        },
      ],
      "name": "use/types",
      "symbols": Array [
        "MeshVertex",
      ],
    },
    Object {
      "imports": Array [
        Object {
          "imported": "viewUniforms",
          "name": "viewUniforms",
        },
        Object {
          "imported": "worldToClip",
          "name": "worldToClip",
        },
      ],
      "name": "use/view",
      "symbols": Array [
        "viewUniforms",
        "worldToClip",
      ],
    },
    Object {
      "imports": Array [
        Object {
          "imported": "getQuadUV",
          "name": "getQuadUV",
        },
      ],
      "name": "geometry/quad",
      "symbols": Array [
        "getQuadUV",
      ],
    },
    Object {
      "exported": true,
      "prototype": Object {
        "name": "getQuadVertex",
        "parameters": Array [
          "int",
          "vertexIndex",
        ],
        "type": Object {
          "name": "MeshVertex",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "getQuadVertex",
      ],
    },
    Object {
      "exported": false,
      "prototype": Object {
        "name": "getPosition",
        "parameters": Array [
          "int",
        ],
        "type": Object {
          "name": "vec4",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "getPosition",
      ],
    },
    Object {
      "exported": false,
      "prototype": Object {
        "name": "getSize",
        "parameters": Array [
          "int",
        ],
        "type": Object {
          "name": "float",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "getSize",
      ],
    },
  ],
  "symbols": Array [
    "MeshVertex",
    "viewUniforms",
    "worldToClip",
    "getQuadUV",
    "getQuadVertex",
    "getPosition",
    "getSize",
  ],
  "visibles": Array [
    "getQuadVertex",
  ],
}
`;

exports[`ast extracts test declarations 1`] = `
Array [
  Object {
    "exported": false,
    "symbols": Array [
      "x",
      "y",
    ],
    "variable": Object {
      "locals": Array [
        Object {
          "expr": null,
          "name": "x",
        },
        Object {
          "expr": null,
          "name": "y",
        },
      ],
      "type": Object {
        "name": "float",
        "qualifiers": Array [],
      },
    },
  },
  Object {
    "exported": false,
    "symbols": Array [
      "a",
      "b",
      "c",
      "d",
    ],
    "variable": Object {
      "locals": Array [
        Object {
          "expr": "3",
          "name": "a",
        },
        Object {
          "expr": "1",
          "name": "b",
        },
        Object {
          "expr": null,
          "name": "c",
        },
        Object {
          "expr": null,
          "name": "d",
        },
      ],
      "type": Object {
        "name": "int",
        "qualifiers": Array [],
      },
    },
  },
]
`;

exports[`ast extracts test declarations with array 1`] = `
Array [
  Object {
    "exported": false,
    "symbols": Array [
      "QUAD",
    ],
    "variable": Object {
      "locals": Array [
        Object {
          "expr": "ivec2(0, 0)",
          "name": "QUAD",
        },
      ],
      "type": Object {
        "name": "ivec2",
        "qualifiers": Array [
          "const",
        ],
      },
    },
  },
]
`;

exports[`ast extracts use types symbol table 1`] = `
Object {
  "declarations": Array [
    Object {
      "exported": true,
      "symbols": Array [
        "MeshVertex",
      ],
      "variable": Object {
        "locals": Array [],
        "type": Object {
          "members": Array [
            Object {
              "name": "position",
              "type": Object {
                "name": "vec4",
              },
            },
            Object {
              "name": "color",
              "type": Object {
                "name": "vec4",
              },
            },
            Object {
              "name": "uv",
              "type": Object {
                "name": "vec2",
              },
            },
          ],
          "name": "MeshVertex",
          "qualifiers": Array [],
        },
      },
    },
  ],
  "externals": Array [],
  "functions": Array [],
  "hash": "y0g6sb9nv5",
  "modules": Array [],
  "refs": Array [
    Object {
      "exported": true,
      "symbols": Array [
        "MeshVertex",
      ],
      "variable": Object {
        "locals": Array [],
        "type": Object {
          "members": Array [
            Object {
              "name": "position",
              "type": Object {
                "name": "vec4",
              },
            },
            Object {
              "name": "color",
              "type": Object {
                "name": "vec4",
              },
            },
            Object {
              "name": "uv",
              "type": Object {
                "name": "vec2",
              },
            },
          ],
          "name": "MeshVertex",
          "qualifiers": Array [],
        },
      },
    },
  ],
  "symbols": Array [
    "MeshVertex",
  ],
  "visibles": Array [
    "MeshVertex",
  ],
}
`;

exports[`ast extracts use view declarations 1`] = `
Array [
  Object {
    "exported": true,
    "struct": Object {
      "name": "viewUniforms",
      "struct": Object {
        "members": Array [
          Object {
            "name": "projectionMatrix",
            "type": Object {
              "name": "mat4",
            },
          },
          Object {
            "name": "viewMatrix",
            "type": Object {
              "name": "mat4",
            },
          },
          Object {
            "name": "viewPosition",
            "type": Object {
              "name": "vec4",
            },
          },
          Object {
            "name": "viewResolution",
            "type": Object {
              "name": "vec2",
            },
          },
          Object {
            "name": "viewSize",
            "type": Object {
              "name": "vec2",
            },
          },
        ],
      },
      "type": Object {
        "name": "ViewUniforms",
        "qualifiers": Array [
          "layout(set = 0, binding = 0)",
          "uniform",
        ],
      },
    },
    "symbols": Array [
      "viewUniforms",
      "ViewUniforms",
    ],
  },
]
`;

exports[`ast extracts use view symbol table 1`] = `
Object {
  "declarations": Array [
    Object {
      "exported": true,
      "struct": Object {
        "name": "viewUniforms",
        "struct": Object {
          "members": Array [
            Object {
              "name": "projectionMatrix",
              "type": Object {
                "name": "mat4",
              },
            },
            Object {
              "name": "viewMatrix",
              "type": Object {
                "name": "mat4",
              },
            },
            Object {
              "name": "viewPosition",
              "type": Object {
                "name": "vec4",
              },
            },
            Object {
              "name": "viewResolution",
              "type": Object {
                "name": "vec2",
              },
            },
            Object {
              "name": "viewSize",
              "type": Object {
                "name": "vec2",
              },
            },
          ],
        },
        "type": Object {
          "name": "ViewUniforms",
          "qualifiers": Array [
            "layout(set = 0, binding = 0)",
            "uniform",
          ],
        },
      },
      "symbols": Array [
        "viewUniforms",
        "ViewUniforms",
      ],
    },
  ],
  "externals": Array [],
  "functions": Array [
    Object {
      "exported": true,
      "prototype": Object {
        "name": "worldToView",
        "parameters": Array [
          "vec4",
          "position",
        ],
        "type": Object {
          "name": "vec4",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "worldToView",
      ],
    },
    Object {
      "exported": true,
      "prototype": Object {
        "name": "viewToClip",
        "parameters": Array [
          "vec4",
          "position",
        ],
        "type": Object {
          "name": "vec4",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "viewToClip",
      ],
    },
    Object {
      "exported": true,
      "prototype": Object {
        "name": "worldToClip",
        "parameters": Array [
          "vec4",
          "position",
        ],
        "type": Object {
          "name": "vec4",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "worldToClip",
      ],
    },
    Object {
      "exported": true,
      "prototype": Object {
        "name": "clipToScreen3D",
        "parameters": Array [
          "vec4",
          "position",
        ],
        "type": Object {
          "name": "vec3",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "clipToScreen3D",
      ],
    },
    Object {
      "exported": true,
      "prototype": Object {
        "name": "screenToClip3D",
        "parameters": Array [
          "vec4",
          "position",
        ],
        "type": Object {
          "name": "vec3",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "screenToClip3D",
      ],
    },
    Object {
      "exported": true,
      "prototype": Object {
        "name": "worldToClip3D",
        "parameters": Array [
          "vec4",
          "position",
        ],
        "type": Object {
          "name": "vec3",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "worldToClip3D",
      ],
    },
  ],
  "hash": "je3ka1b7aj",
  "modules": Array [],
  "refs": Array [
    Object {
      "exported": true,
      "prototype": Object {
        "name": "worldToView",
        "parameters": Array [
          "vec4",
          "position",
        ],
        "type": Object {
          "name": "vec4",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "worldToView",
      ],
    },
    Object {
      "exported": true,
      "prototype": Object {
        "name": "viewToClip",
        "parameters": Array [
          "vec4",
          "position",
        ],
        "type": Object {
          "name": "vec4",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "viewToClip",
      ],
    },
    Object {
      "exported": true,
      "prototype": Object {
        "name": "worldToClip",
        "parameters": Array [
          "vec4",
          "position",
        ],
        "type": Object {
          "name": "vec4",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "worldToClip",
      ],
    },
    Object {
      "exported": true,
      "prototype": Object {
        "name": "clipToScreen3D",
        "parameters": Array [
          "vec4",
          "position",
        ],
        "type": Object {
          "name": "vec3",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "clipToScreen3D",
      ],
    },
    Object {
      "exported": true,
      "prototype": Object {
        "name": "screenToClip3D",
        "parameters": Array [
          "vec4",
          "position",
        ],
        "type": Object {
          "name": "vec3",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "screenToClip3D",
      ],
    },
    Object {
      "exported": true,
      "prototype": Object {
        "name": "worldToClip3D",
        "parameters": Array [
          "vec4",
          "position",
        ],
        "type": Object {
          "name": "vec3",
          "qualifiers": Array [],
        },
      },
      "symbols": Array [
        "worldToClip3D",
      ],
    },
    Object {
      "exported": true,
      "struct": Object {
        "name": "viewUniforms",
        "struct": Object {
          "members": Array [
            Object {
              "name": "projectionMatrix",
              "type": Object {
                "name": "mat4",
              },
            },
            Object {
              "name": "viewMatrix",
              "type": Object {
                "name": "mat4",
              },
            },
            Object {
              "name": "viewPosition",
              "type": Object {
                "name": "vec4",
              },
            },
            Object {
              "name": "viewResolution",
              "type": Object {
                "name": "vec2",
              },
            },
            Object {
              "name": "viewSize",
              "type": Object {
                "name": "vec2",
              },
            },
          ],
        },
        "type": Object {
          "name": "ViewUniforms",
          "qualifiers": Array [
            "layout(set = 0, binding = 0)",
            "uniform",
          ],
        },
      },
      "symbols": Array [
        "viewUniforms",
        "ViewUniforms",
      ],
    },
  ],
  "symbols": Array [
    "worldToView",
    "viewToClip",
    "worldToClip",
    "clipToScreen3D",
    "screenToClip3D",
    "worldToClip3D",
    "viewUniforms",
    "ViewUniforms",
  ],
  "visibles": Array [
    "worldToView",
    "viewToClip",
    "worldToClip",
    "clipToScreen3D",
    "screenToClip3D",
    "worldToClip3D",
    "viewUniforms",
    "ViewUniforms",
  ],
}
`;

exports[`ast rewrites code using the AST 1`] = `
"
    float _zz_getValue(int index);
    void entryPoint() {
      float x = 3.0;
      float y = _zz_getValue(2);
    }
    "
`;

exports[`ast rewrites code using the compressed AST 1`] = `
Array [
  Array [
    "Identifier",
    11,
    19,
  ],
  Array [
    "Identifier",
    24,
    29,
  ],
  Array [
    "Identifier",
    41,
    45,
  ],
  Array [
    "Identifier",
    62,
    63,
  ],
  Array [
    "Identifier",
    83,
    84,
  ],
  Array [
    "Identifier",
    87,
    95,
  ],
  Array [
    "Identifier",
    111,
    112,
  ],
  Array [
    "Identifier",
    120,
    121,
  ],
]
`;

exports[`ast rewrites code using the compressed AST 2`] = `
Object {
  "cursor": [Function],
}
`;
